src/conn.ts

```diff
@@ -1,183 +1,376 @@
+   * @param pclient Client
+   */
+  _serverClientDefaultMiddleware(pclient: Client) {
+    if (!pclient.toClientMiddlewares) pclient.toClientMiddlewares = [];
+    const _internalMcProxyServerClient: PacketMiddleware = (info, pclient, data, cancel, update) => {
+      if (!this.receivingClients.includes(pclient)) return cancel();
+    };
+    pclient.toClientMiddlewares.push(_internalMcProxyServerClient);
+    if (this.toClientDefaultMiddleware) pclient.toClientMiddlewares.push(...this.toClientDefaultMiddleware);
+  }
+
+  /**
+   * Register the default (first) middleware used to control what client can interact with the current bot.
+   * @param pclient Client
+   */
+  _clientServerDefaultMiddleware(pclient: Client) {
+    if (!pclient.toServerMiddlewares) pclient.toServerMiddlewares = [];
+    const _internalMcProxyClientServer: PacketMiddleware = (info, pclient, data: any, cancel) => {
+      if (info.meta.state !== 'play') return cancel();
+      if (info.meta.name === 'teleport_confirm' && data?.teleportId === 0) {
+        pclient.write('position', {
+          ...this.bot.entity.position,
+          yaw: 180 - (this.bot.entity.yaw * 180) / Math.PI,
+          pitch: -(this.bot.entity.pitch * 180) / Math.PI,
+          teleportId: 1,
+        });
+        cancel();
+      }
+      //* check if client is authorized to modify connection (sending packets and state information from mineflayer)
+      if (this.writingClient !== pclient) {
+        return cancel();
+      }
+      // Keep the bot updated
+      // Note: Packets seam to be the exact same going from server to client and the other way around.
+      // At least for 1.12.2. So this is just copy past from onServerRaw
+      switch (info.meta.name) {
+        case 'position':
+          this.bot.entity.position.x = data.x;
+          this.bot.entity.position.y = data.y;
+          this.bot.entity.position.z = data.z;
+          this.bot.entity.onGround = data.onGround;
+          this.bot.emit('move', this.bot.entity.position) // If bot is not in control physics are turned off
+          break;
+        case 'position_look': // FALLTHROUGH
+          this.bot.entity.position.x = data.x;
+          this.bot.entity.position.y = data.y;
+          this.bot.entity.position.z = data.z;
+        case 'look':
+          this.bot.entity.yaw = ((180 - data.yaw) * Math.PI) / 180;
+          this.bot.entity.pitch = -(data.pitch * Math.PI) / 180;
+          this.bot.entity.onGround = data.onGround;
+          this.bot.emit('move', this.bot.entity.position) // If bot is not in control physics are turned off
+          break;
+        case 'held_item_slot':
+          this.bot.quickBarSlot = data.slot;
+          this.bot._client.emit('mcproxy:heldItemSlotUpdate') // lol idk how to do it better
           break;
         case 'abilities':
-          this.bot.physicsEnabled = !!((data.flags & 0b10) ^ 0b10);
+          this.bot.physicsEnabled = !this.writingClient && !!((data.flags & 0b10) ^ 0b10);
```

Comment:
why this change?

Reply:
To disable the bots physics when the bot is not in control to save on performance and prevent the bot from having weird side effects while not being in control off the connection